### PR TITLE
Add flexible rate option support for graph generation

### PR DESCRIPTION
**Pull request by jshaughn**
Supports https://github.com/kiali/kiali/issues/3730
Supports https://github.com/kiali/kiali/issues/4070

**UI PR:** https://github.com/kiali/kiali-ui/pull/2181
**OPERATOR PR:** https://github.com/kiali/kiali-operator/pull/342

### CODE REVIEW

File: graph/options.go [hunk 1]
```diff
@@ -33,11 +33,19 @@ const (
...
+	RateRequests              string = "requests" // request count
+	RateSent                  string = "sent"     // tcp bytes sent, grpc request messages, etc
+	RateTotal                 string = "total"    // Sent+Received
 	defaultBoxBy              string = BoxByNone
 	defaultDuration           string = "10m"
 	defaultGraphType          string = GraphTypeWorkload
 	defaultIncludeIdleEdges   bool   = false
 	defaultInjectServiceNodes bool   = false
+	defaultRateHttp           string = RateRequests
+	defaultRateGrpc           string = RateRequests
```
Comment (jmazzitelli) [inline graph/options.go:47]: RED ALERT!!!! What blasphemous code doth my eyes witness?

```suggestion
	defaultRateGrpc           string = RateRequests
	defaultRateHttp           string = RateRequests
```

**Review by jmazzitelli — COMMENTED** [review]
Aside from the egregious non-alphabetized order of some of this code, I built and ran this and all looks good.

File: graph/options.go (hunk 1 above)
Comment (jshaughn) [reply graph/options.go:47]: O M G, will fix.

**Review by jmazzitelli — DISMISSED** [review]
LGTM

**Comment by lucasponce** [comment]
I see an increasing concern to desconect the story that the user wants to see.

If I want to generate a topology from "GRPC received messages" when the user would like to examine a chart in detail, that metrics won't be available as this is fixed by requests.

The original story of the details metrics was to give the user the ability to get more details and combine with other information.

I think that story should be keep, and probably the details could be organized by protocols (grpc, http, tcp) or another combination that follows the same story.

I haven't found anything broken in that sense, a couple of problems due performance, but I need to repeat some tests tomorrow to discard that it has nothing to do with the PR.

I'll take a look close to the code, in a first pass I focused on the feature.

**Comment by jshaughn** [comment]
> I see an increasing concern to desconect the story that the user wants to see.
If I want to generate a topology from "GRPC received messages" when the user would like to examine a chart in detail, that metrics won't be available as this is fixed by requests.
The original story of the details metrics was to give the user the ability to get more details and combine with other information.I think that story should be keep,

I agree and I think the question is how much change to introduce incrementally.  The need to better handle different traffic types has been growing, first came TCP issues when sentBytes was unable to visualize some of the TCP traffic.  And now the gRPC streaming problem.  If we introduce this into the main graph users will expect the support to extend in some logical way to potentially relevant areas like side-panel, mini-graph, charts.   

One thing that may not be able to wait for a follow-up PR is some work in the side-panel.   As it stands now if you select a gRPC or TCP edge and you have selected a non-default rate, the data doesn't really match up very well.  Let me know what you think about breaking up the work or adding it to this initial PR.

> ... and probably the details could be organized by protocols (grpc, http, tcp) or another combination that follows the same story.

Note that gRPC is a pita, because it offers both request-based rates OR streaming-type rates (based on message counts), so once protocol can still force different metrics based on the desired rate.

> I haven't found anything broken in that sense, a couple of problems due performance, but I need to repeat some tests tomorrow to discard that it has nothing to do with the PR.

Note that certain options will increase graph-gen time.  As you may expect, more queries are required depending on what is being asked for.  On the bright side, eliminating a protocol, or requests,  may speed things up.

I'll take a look close to the code, in a first pass I focused on the feature.

**Comment by jshaughn** [comment]
UPDATE Jul 1:  Looking more at the side-panel for a gRPC edge using message rate, we need to fix the side panel as part of this work.  So, I will work on that now...

**Comment by lucasponce** [comment]
This one would require to rebase with master now that https://github.com/kiali/kiali/pull/4171 is merged.

**Comment by jshaughn** [comment]
@lucasponce The server and UI PRs are now rebased with your #4171 and any other recent changes.

**Comment by pbajjuri20** [comment]
Tested and verified the TCP, HTTP and GRPC traffic - 1.38.0 (4bdd7cb40fe80e856a62250ac1830a67d770b9bd)

**Review by pbajjuri20 — APPROVED** [review]

**Review by jmazzitelli — APPROVED** [review]
in praneenth I trust.
lgtm

**Pull request merged**